src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/DoNotCallOrderByMultipleTimes.cs

```diff
@@ -0,0 +1,85 @@
+
+namespace Microsoft.NetCore.Analyzers.Usage
+{
+    using static MicrosoftNetCoreAnalyzersResources;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public class DoNotCallOrderByMultipleTimes : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2259";
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(DoNotCallOrderByMultipleTimesTitle)),
+            CreateLocalizableResourceString(nameof(DoNotCallOrderByMultipleTimesDescription)),
+            DiagnosticCategory.Usage,
+            RuleLevel.BuildWarning,
+            description: null,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+            ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var compilation = context.Compilation;
+
+            if (!compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemLinqEnumerable, out var enumerableType))
+                return;
+
+            var analyzeOperation = new AnalyzeOperation(enumerableType);
+
+            context.RegisterOperationAction(ctx => analyzeOperation.Analyze(ctx),
+                                            OperationKind.Invocation);
+        }
+
+        private class AnalyzeOperation
+        {
+            public INamedTypeSymbol EnumerableType { get; }
+            public IInvocationOperation? PreviousOrderByInvocation { get; private set; }
+
+            public AnalyzeOperation(INamedTypeSymbol enumerableType)
+            {
+                EnumerableType = enumerableType;
+            }
+
+            public void Analyze(OperationAnalysisContext context)
+            {
+                if (context.Operation is not IInvocationOperation invocationOperation)
+                    return;
+
+                if (invocationOperation.TargetMethod.ReceiverType != EnumerableType)
+                    return;
+
+                if (invocationOperation.TargetMethod.Name is not "OrderBy" and not "OrderByDescending")
+                    return;
+
+                if (PreviousOrderByInvocation is null)
+                {
+                    PreviousOrderByInvocation = invocationOperation;
+                }
+                else if (PreviousOrderByInvocation != invocationOperation)
+                {
+                    context.ReportDiagnostic(PreviousOrderByInvocation.CreateDiagnostic(Rule));
+                }
```

Comment:
I'm feeling this will incorrectly trigger a diagnostic for:


```csharp
var x = MyCollection.OrderBy(c => c.X);
var y = MyOtherCollection.OrderBy(c => c.Y);
```

Reply:
Probably. My other concern is a false negative:

```csharp
var x = MyCollection.OrderBy(c => c.X);

var y = x.OrderBy(c => c.Y); // this should report the diagnostic
```

I tried (unsuccessfully, so far) to instead see if the type of `x` here is `IOrderedEnumerable`.